### PR TITLE
Add pdm lockfile support

### PR DESCRIPTION
**Pull request by jtt**
Add support for parsing package information from  `pdm.lock` -files used by `pdm`, package and dependency manager for Python (https://pdm-project.org/latest/)

### CODE REVIEW

File: pkg/lockfile/parse-pdm-lock.go
Comment (G-Rath) [inline pkg/lockfile/parse-pdm-lock.go:1]: would you mind restructuring these to use individual functions to be consistent with the rest of our lockfile tests

File: pkg/lockfile/parse-pdm-lock.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+type PdmLockFile struct {
+	Version  string           `toml:"lock-version"`
+	Packages []PdmLockPackage `toml:"package"`
+}
+
+const PdmEcosystem = PipEcosystem
+const PdmLockFileName = "pdm.lock"
+const pdmGroupDefault = "default"
+const pdmGroupDev = "dev"
```
Comment (G-Rath) [inline pkg/lockfile/parse-pdm-lock.go:25]: we don't typically use constants for these, and you're only referencing them each once so lets just inline them

File: pkg/lockfile/parse-pdm-lock.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	"fmt"
+	"path/filepath"
+
+	"github.com/BurntSushi/toml"
+)
+
+type PdmLockPackage struct {
+	Name     string   `toml:"name"`
+	Version  string   `toml:"version"`
+	Groups   []string `toml:"groups"`
```
Comment (G-Rath) [inline pkg/lockfile/parse-pdm-lock.go:13]: just noting for others: it looks like pdm only _very_ recently started recording dependency groups at the package level in its lockfiles - I couldn't actually find any locks searching GitHub, but the PRs on pdm support this feature as being added

File: pkg/lockfile/parse-pdm-lock.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+			details.Commit = pkg.Revision
+		}
+
+		packages = append(packages, details)
+	}
+
+	return packages, nil
+}
+
+//nolint:gochecknoinits
```
Comment (G-Rath) [inline pkg/lockfile/parse-pdm-lock.go:73]: we should ensure the extractor matches the interface:

```suggestion
}

var _ Extractor = PdmLockExtractor{}

//nolint:gochecknoinits
```

File: pkg/lockfile/parse-pdm-lock.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	Version  string           `toml:"lock-version"`
+	Packages []PdmLockPackage `toml:"package"`
+}
+
+const PdmEcosystem = PipEcosystem
+const PdmLockFileName = "pdm.lock"
+const pdmGroupDefault = "default"
+const pdmGroupDev = "dev"
+
+type PdmExtractor struct{}
```
Comment (G-Rath) [inline pkg/lockfile/parse-pdm-lock.go:27]: This matches the naming we use for the other extractors:

```suggestion
type PdmLockExtractor struct{}
```

**Review by G-Rath — CHANGES_REQUESTED** [review]
Looks good, just got a few changes - would you mind also adding an "empty.lock" (which is "no packages", not "completely empty file") and "two-packages.lock" fixtures as well? (I know technically your other tests cover this, but it's consistent with the standard set of fixtures we have for our other lockfiles)

File: pkg/lockfile/parse-pdm-lock.go
Comment (jtt) [reply pkg/lockfile/parse-pdm-lock.go:1]: Restructured

File: pkg/lockfile/parse-pdm-lock.go (hunk 1 above)
Comment (jtt) [reply pkg/lockfile/parse-pdm-lock.go]: Removed constants

File: pkg/lockfile/parse-pdm-lock.go (hunk 3 above)
Comment (jtt) [reply pkg/lockfile/parse-pdm-lock.go]: Added.

File: pkg/lockfile/parse-pdm-lock.go (hunk 4 above)
Comment (jtt) [reply pkg/lockfile/parse-pdm-lock.go]: Changed the name

**Comment by jtt** [comment]
> Looks good, just got a few changes - would you mind also adding an "empty.lock" (which is "no packages", not "completely empty file") and "two-packages.lock" fixtures as well? (I know technically your other tests cover this, but it's consistent with the standard set of fixtures we have for our other lockfiles)

Added requested tests. Amended fixes to the existing commit.

File: pkg/lockfile/parse-pdm-lock_test.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+func expectNilErr(t *testing.T, err error) {
+	t.Helper()
+	if err != nil {
+		t.Errorf("got unexpected error: %v", err)
+	}
+}
+
+func TestParsePdmLock_FileDoesNotExist(t *testing.T) {
+	t.Parallel()
```
Comment (G-Rath) [inline pkg/lockfile/parse-pdm-lock_test.go:76]: nit: these are pretty squashed together

usually I like to have a new line between every function, after `t.Parallel()`, and before the first `expect*` call

**Review by G-Rath — APPROVED** [review]
Awesome stuff! It'd be great if you could space out the tests a bit, but I think this is good to go!

File: pkg/lockfile/parse-pdm-lock_test.go (hunk 5 above)
Comment (jtt) [reply pkg/lockfile/parse-pdm-lock_test.go:76]: Added newlines to test functions.

**Review by another-rex — APPROVED** [review]
Thanks!

**Pull request merged**